qiskit/opflow/gradients/circuit_gradients/param_shift.py

```diff
@@ -205,17 +205,22 @@ def _parameter_shift(
 
             shifted_ops = []
             summed_shifted_op = None
-            for m, param_occurence in enumerate(circ._parameter_table[param]):
-                param_index = param_occurence[1]
+
+            iref_to_data_index = {id(inst): idx for idx, (inst, _, _) in enumerate(circ.data)}
+
+            for param_reference in circ._parameter_table[param]:
+                original_gate, param_index = param_reference
+                m = iref_to_data_index[id(original_gate)]
+
                 pshift_op = deepcopy(operator)
                 mshift_op = deepcopy(operator)
 
                 # We need the circuit objects of the newly instantiated operators
                 pshift_circ = self.get_unique_circuits(pshift_op)[0]
                 mshift_circ = self.get_unique_circuits(mshift_op)[0]
 
-                pshift_gate = pshift_circ._parameter_table[param][m][0]
-                mshift_gate = mshift_circ._parameter_table[param][m][0]
+                pshift_gate = pshift_circ.data[m][0]
+                mshift_gate = mshift_circ.data[m][0]
```

Comment:
Yeah... I don't have a complete understanding of what this code is doing. It doesn't seem like it takes into account that an instruction might reference multiple parameters (and hence multiple parameter reference entries might contain the same instruction). I tried to stay as close to the original logic to avoid regressions.